File: day5/part1.py
#!/usr/bin/env python3

import sys

def main(): 
    if len(sys.argv) != 2: 
        print(f"[usage]: {sys.argv[0]} [puzzle.txt]", file=sys.stderr)

    seeds = []
    maps = []

    # parse file
    with open(sys.argv[1], "r") as f: 
        state = -1
        started = False

        for line in f: 
            if state == -1: 
                if len(line) == 1: 
                    state += 1
                s = line.split(" ")[1:]
                for n in s: 
                    seeds.append(int(n))

            else: 
                if started == False: 
                    started = True 
                    maps.append([])
                    continue 
                elif started == True: 
                    if len(line) == 1: 
                        state += 1
                        started = False 
                    else: 
                        s = line.split(" ")
                        res = list(map(int, s))
                        maps[state].append(res)

    loc = []
    for s in seeds: 
        curr = s
        for m in maps: 
            for l in m: 
                if curr >= l[1] and curr < l[1] + l[2]: 
                    offset = curr - l[1]
                    curr = l[0] + offset
                    break
        loc.append(curr)

    print(min(loc))



if __name__ == "__main__": 
    main()

File: day5/part1.go
package main

import (
    "os"
    "io"
    "log"
    "fmt"
    "strconv"
    "strings"
)

func main() {
    if len(os.Args) != 2 {
        fmt.Fprintf(os.Stderr, "[usage]: %s [puzzle]\n", os.Args[0])
        os.Exit(1)
    }

    var seeds []int64
    var maps [][][3]int64

    // parse file
    file, err := os.Open(os.Args[1])
    if err != nil {
        log.Fatal(err)
    }

    f, err := io.ReadAll(file)
    if err != nil {
        log.Fatal(err)
    }
    file.Close()

    lines := strings.Split(string(f), "\n")

    state := -1
    started := false

    for _, line := range lines {
        if state == -1 {
            if len(line) == 0 {
                state += 1
            } else {
                s := strings.Split(line, " ")[1:]
                for _, n := range s {
                    res, err := strconv.ParseInt(n, 10, 64)
                    if err != nil {
                        log.Fatal(err)
                    }
                    seeds = append(seeds, res)
                }
            }
        } else {
            if started == false {
                started = true
                maps = append(maps, [][3]int64{})
                continue
            } else {
                if len(line) == 0 {
                    state += 1
                    started = false
                } else {
                    s := strings.Split(line, " ")
                    n1, err := strconv.ParseInt(s[0], 10, 64)
                    if err != nil {
                        log.Fatal(err)
                    }
                    n2, err := strconv.ParseInt(s[1], 10, 64)
                    if err != nil {
                        log.Fatal(err)
                    }
                    n3, err := strconv.ParseInt(s[2], 10, 64)
                    if err != nil {
                        log.Fatal(err)
                    }
                    maps[state] = append(maps[state], [3]int64{n1, n2, n3})
                }
            }
        }
    }

    var min_num int64 = 9999999999
    for _, s := range seeds {
        curr := s
        for _, m := range maps {
            for _, l := range m {
                if curr >= l[1] && curr < l[1] + l[2] {
                    offset := curr - l[1]
                    curr = l[0] + offset
                    break
                }
            }
        }
        if curr < min_num {
            min_num = curr
        }
    }

    fmt.Println(min_num)
}
